### ISSUE TITLE
Bokeh server questions

### ISSUE THREAD

**Issue by birdsarah** [bokeh/bokeh#3279]
We've already discussed this a little, and a request came up today with @stefanv and @acrellin considering bokeh for https://github.com/mltsp/mltsp.

Need / question (1):
- web application has multiple logged in users
- the users have their own dataset
- they want to access a common plot  e.g.`nd_scatter`
- **the plot `nd_scatter` should be populated with the users data.**
- **the users should not be able to see other users data** 

Need / question (2):
- web application has multiple logged in users
- the users have their own dataset
- they site has multiple plots
- **the user should only be able to access plots that they have access to**

Need / question (3):
- the application has 10 separate Documents - each Document has 1 - 3 connected plots.
- the Documents are unique to different web pages on a website
- **how do you start bokeh server to start serving all these Documents**
- **what do you run in the web application code to get the script/div for embedding a given Document in a new template (in a standalone context I would use `embed.components`)**

The answers to these questions may not result in additions to bokeh, but simply recommendations on deployment setups.

General thoughts:
- I don't think security through obfuscation (long uid) is sufficient - we should support/enable auth (I don't mean implement just have enough so it can be done)
- We should consider servers living behind proxies and firewalls etc

**Comment by stefanv** [comment]
Tying in to this may be a discussion on how to deploy a Bokeh server when you already have nginx, flask/tornado/webserver-xyz and websockets in the mix.  Our deployment situation is illustrated here, e.g.:

https://camo.githubusercontent.com/0b87751556afce5f57de01ae59bfc15d974b81fb/687474703a2f2f7777772e676c696666792e636f6d2f676f2f7075626c6973682f696d6167652f393139303831392f4c2e706e67

**Comment by bryevdv** [comment]
So just to share my own plans / what I intend to have for the release docs:

* example of Bokeh server behind nginx (with or without simple auth) This has pretty much been done and just needs to be written up in a nice form. 
* example of a bokeh server extended with additional views (so, tornado-based web app extending bokeh server directly) This should be very easy to put together
* example of using bokeh server alongside a flask (or django, whatever) app. I expect it will largely be the flask app serving up `autoload_server` scripts that can embed the plots from the bokeh server (so we will have to advise on nginx configuration, etc) This is the one that will require a little more thought and work. And to prevent unauthorized viewing we might have to implement some kind of simple token scheme so that only views that went through the flask app and have the correct token get served the plot from bokeh server. 

ping @havocp

**Comment by havocp** [comment]
(typed this up before I saw Bryan's comment above)

I think there's a lot here that still needs figuring out. My view is that Bokeh can't/shouldn't provide an "account system," only hooks needed to integrate Bokeh into a larger picture that might include such a system. I would lean toward hooks that are extremely general (like filters on incoming requests / outgoing responses) rather than something like a "directory server API," if possible.

Ideas for ways auth could work in a given deployment:

 * put it in the reverse proxy (nginx or whatever proxy you have in front of bokeh). I don't know that nginx supports this beyond http basic auth, but maybe some proxies do or could. Or people could even write a reverse proxy using Tornado (in a quick google, this has been done).
 * put it in the app (meaning the bokeh app - the script itself would present a password textbox, then store the user's identity in the session document, and use that identity in callbacks etc.)
 * put it in the app (meaning a web app built with flask/django/etc. that embeds bokeh plots); if using autoload_server, we'd need some way for the flask/django/whatever app to tell bokeh the request is authorized. (Perhaps we could invent a system where bokeh serve and the flask/django app have a shared secret, and the browser has to provide a hash of the secret to the bokeh server. This could look like `bokeh serve --auth-secret=foo` and then autoload_server would take the secret as a parameter, perhaps, not sure.)
 * support adding some sort of filter on the tornado level (let people pre-process requests or post-process responses) ?

It may well be that Bokeh should support all of the above, which is why we didn't want to build auth in, because there are so many ways a deployment might need to work.

Related but separate, we should ensure that bokeh serve sessions are secure (meaning the session id is secret and cryptographically random). One piece of work related to this would be to have a variant of pull_session() that lets the server generate the session ID, and potentially a configuration of the server that only allows server-generated session IDs.

This thread is kind of related: https://groups.google.com/a/continuum.io/forum/#!searchin/bokeh/havoc/bokeh/xrk240aPu88/QR6aEDELCgAJ
the idea being that the flask app would ask bokeh serve for a new session, and then use that new session id for the autoload_server, but the flask app shouldn't need to actually pull the entire Document into the flask process. If we had a configuration of `bokeh serve` that would only accept sessions it had generated, and if we had a shared secret so that only the flask app was allowed to create a new session, that could work well. 

Also the Flask app could use the client API to customize each session for example inserting different data for each user.

That may be the easiest/shortest path to something that works nicely with Flask: run `bokeh serve` with empty apps, and have the Flask app create sessions and upload the document for each session, then pass the session id down to autoload_server. Then the Flask app could do all the auth and be sure the right thing for each user is in that user's session.

**Comment by havocp** [comment]
so, less rambly... I'm liking the idea of the Flask (or Django, whatever, Flask is shorthand for "some web framework") app just treating `bokeh serve` as a session repository. This I think would support all of Sarah's scenarios.

Here's how it works. You run `bokeh serve` with no parameters, there's one application which is empty.

Each time you want to embed a plot, the logic is all in the Flask app. The Flask app does a `push_session` to create a new session, and puts whatever Document it wants into that session. It then uses the session id for the session to autoload_server and generate HTML.

One user could have multiple sessions, and even though on the `bokeh serve` side we have one application, really the different sessions need not have anything in common, they can be totally different Document.

The one problem here is that there's no way to use `push_session` to upload interactive code, so at that point there's really no point having Bokeh server at all, if you can only do standalone documents you might as well have just generated standalone html ... now, we COULD run callbacks in the Flask app, by keeping the Flask-side copy of the Document around and holding the session open, but this will get complicated since the Flask app would now have to manage an IO loop for all those sessions.

I don't know the solution to this yet. Some way to upload code... ?

Perhaps that's a fatal problem with the idea, so potentially we do need a separate Bokeh app for each document type - things could still be structured similarly where the Flask app manages sessions, but sessions would be pre-initialized with some `bokeh serve`-side code, and Flask would only modify the document rather than creating it from scratch.

**Comment by bryevdv** [comment]
>Also the Flask app could use the client API to customize each session for example inserting different data for each user.

> That may be the easiest/shortest path to something that works nicely with Flask: run bokeh serve with empty apps, and have the Flask app create sessions and upload the document for each session, then pass the session id down to autoload_server. Then the Flask app could do all the auth and be sure the right thing for each user is in that user's session.

I think it can work fine with `bokeh serve app.py` too? The client API can connect to request a new session and then that session can be "forwarded" on with `autoload_server`. The ability to add customizations via the client API, or start from empty docs and build up from there is also nice, but mostly in addition to this interaction pattern I think

**Comment by birdsarah** [comment]
> The one problem here is that there's no way to use push_session to upload interactive code, so at that point there's really no point having Bokeh server at all, if you can only do standalone documents you might as well have just generated standalone html 

Right. That's kind of the heart of the issue. Making a static site with bokeh is a solved problem.

> I think it can work fine with bokeh serve app.py too? The client API can connect to request a new session and then that session can be "forwarded" on with autoload_server.

I don't know why this isn't obvious to me but I am really struggling with exactly how this works.

I run `bokeh serve foo.py bar.py fizz.py buzz.py` so now bokeh server is capable of serving sessions of four different Documents (feel free to correct my terminology).

On my page foo.html I want to embed the Document that `foo.py` makes. What exactly do I pass to `autoload_server` to get the right thing back?

**Comment by bryevdv** [comment]
Can we have a call? I could spend the next three days typing otherwise. (I'm already making a dr appt about sudden arm/wrist pain)

**Comment by havocp** [comment]
oh also @birdsarah , as I was just bringing up on flowdock - the default session_id for `autoload_server` right now is BAD (it's `'default'` which will put all users in the same non-secret sesssion). So we have to fix that. For now be sure to do `autoload_sever(app_path='/foo', session_id=None`)`

**Comment by birdsarah** [comment]
Just had a call with @bryevdv and @havocp 

What we can do now:

Your webapp lives at: www.allthedata.com
Your bokeh server at: bokeh.allthedata.com

Start bokeh server with all your different app code `bokeh serve foo.py bar.py fizz.py buzz.py`

Now, anyone can go to http://bokeh.allthedata.com/foo, http://bokeh.allthedata.com/bar, http://bokeh.allthedata.com/fizz, and http://bokeh.allthedata.com/buzz

And they will get a session of bokeh server back. We have a new TODO to make it so new sessions can only be created by a trusted party. However, as a short-term workaround you can make that session very empty and let your webapp populate it.

Next step in your django/flask app, authenticate your user, do whatever setup is necessary.

Now, ask bokeh server for a session `pull_session(url='foo')` this is the same as the web path noted above. (There is another discussion about making that path configurable).

Now, push state / data into that session e.g. 

```python
session = pull_session(url='/foo')
document = session.document
source = session.document.source  # Note that you should have added the columndatasource to the document so you can update it
source.data = my_django_users_data  # which you've previously generated
```
Behind the scenes in your foo.py, bokeh will now know what to do when the data updates.

You don't have to get the data in your django app, you could just pass bokeh server your credentials and have the bokeh app code get the data for you. But the basic idea is the same, and will give you a custom session for your user that you can update with whatever data you want.

Next, we serve that session to the user.

```python
script, div = autoload_server(app_path='/foo', session_id=session.id)
# continue as usual passing script/div into your template renderer.
```

**Comment by birdsarah** [comment]
On a seperate thread of discussion, we discussed having bokeh-server data updates be event driven - so you could call to bokeh-server and say, my data updated. This is currently not possible and is tricky so won't be for a while.

**Comment by havocp** [comment]
thanks for the writeup @birdsarah, flawless. (and better than my rambling above.)

I thought of a related topic, which is https://github.com/bokeh/bokeh/issues/3102 - when you have bokeh.allthedata.com, you'd have to `--allow-origin *.allthedata.com` or something along those lines, once we fix things so they aren't wide open.

**Comment by birdsarah** [comment]
@havocp just to clarify: right now, you don't need the `--allow-origin` and bokeh server allows all, but we're closing that loop.

**Comment by birdsarah** [comment]
@havocp - I'm looking at #3171 where you say "It's a pain to construct / figure out the websocket url for pull_session"

If my instructions above are correct why, is it a pain. If they are not, can you provide me the precise string I would pass to `pull_session`?

**Comment by havocp** [comment]
Precisely, right now it's wide open, but we have to fix it.

**Comment by havocp** [comment]
ah, yeah I didn't notice that url issue. `pull_session` wants a thing like `ws://bokeh.allthedata.com/foo/ws`. So #3171 is that it should be maybe `pull_session(url='http://bokeh.allthedata.com', app_path='/foo')` or along those lines, and then it assembles it for you.

**Comment by birdsarah** [comment]
updated above

**Comment by birdsarah** [comment]
@bryevdv @havocp. New question: what's the difference between using curdoc() and Document() to set up a new app?

**Comment by havocp** [comment]
in the script passed to "bokeh serve" you must use curdoc() because it's the Document that will be passed to the server side session.

When talking to bokeh server from another process, curdoc() is the global document used by the bokeh.io convenience API. If you're using pull_session you could generally ignore curdoc().  If you're using push_session you could use and push curdoc() if desired, but since it's a global you'd have to be very careful to only attach one session to it at a time or things might get weird.

**Comment by birdsarah** [comment]
Got it - thanks!

Another new question - lets say I have app code that looks like this:

```python
# foo.py
doc = curdoc()

params = dict(userid=None, params=None, auth_token=None)

def update_me():
    data = do_a_rest_call(params)
    source.data = data

doc.add_periodic_callback(update_me, 100)
```

What I want to do in the code that's going to work with this doc is pass in my new params.

Previously I have been able to set data by doing `pull_session`, then `session.document.get_model_by_name('source')` which returns my source and then I can update the source. Now I could make params a ColumnDataSource just to make this work, but that seems a little silly when I just want to be able to pass in some arbitrary object.  

Am I right that making a `ColumnDataSource` to hold this data is the only way of working like this at the moment

**Comment by havocp** [comment]
You could probably make a custom model to hold the info, but that is ugly and I'd have to go figure out what it looks like (though @brendancol has been doing it a lot lately and may know right away). I think this is why Bryan wants to add Document.vars which would be a Dict(String,Any) auto-synced between client and server for exactly these kind of quick little pieces of data. Not sure there's a good answer right now.

**Comment by birdsarah** [comment]
OK Cool. Interestingly, in this particular case, I don't particularly want them synced between the client and the server. So maybe we have `private_vars` too. 

(@bryevdv is there an issue for Document.vars already I couldn't see one?)

**Comment by havocp** [comment]
I guess here you'd want them synced between your Django app (which is a client of bokeh server) and bokeh server, but not between the browser (which is also a client of bokeh server) and bokeh server.

That could get weird.

First brainstorm idea: some kind of "session create vars" that are available during the initial session creation script here, but not synced with the Document itself. Would presumably be passed to pull_session or push_session.

**Comment by havocp** [comment]
Should probably make an issue. I didn't see one for vars in a quick search.

**Comment by birdsarah** [comment]
@havocp @bryevdv - another new server question :D

See: https://github.com/bokeh/bokeh-demos/blob/7764c281797760ef37ac87054320bc8f5e0fc041/happiness/viz/individual.py#L45

In my periodic_callback, I do: `source.data = new_data`

But what I want to happen is that an update goes down to the client only when there's new data. But I'm having a hard time making that happen and I would appreciate suggestions of how best to do this.

Part of the problem is that data comes out of django in `datetime.date` and `Decimal` but by the time its been eaten by a ColumnDataSource it's all converted into the types that go down the pipe. My current idea is to take the new_data; put it into a ColumnDataSource; do the to_json; then do a compare. But that seems crazy.

**Comment by havocp** [comment]
hmm. I'm not sure of the details here - I think ideally ColumnDataSource would "ingest" the django values when they are assigned, rather than after a roundtrip through JSON - it's really kind of a bug anytime roundtrip through JSON changes a model.

If the only way to get the values in final form is to encode as json, I don't see much alternative to your to_json approach. Possibly you could do the conversion yourself from datetime/Decimal?

**Comment by birdsarah** [comment]
> Possibly you could do the conversion yourself from datetime/Decimal?
I could; but I'm more confident its consistent if I just run to_json. 
Thanks!

**Comment by birdsarah** [comment]
Another new question: is it possible to generate a new session from a static webpage?

Scenario - I have a one page static site: http://mybokehdashboard.com

It's just handcoded html, but in it is a bokeh script tag, that is going to pull in a bokeh server plot. I want every new user who comes to my site to get a unique session. Is it possible to do this with static html; or do I have to do this with a one page flask app that generates a new session id?

**Comment by havocp** [comment]
Short answer no :-) long answer: https://github.com/bokeh/bokeh/issues/3355

**Comment by ppatel23** [comment]
Is the write up by @birdsarah still the best way to set up a bokeh server with data being determined/extracted at runtime? 

I have a setup very similar to @birdsarah on heroku (two apps, one running a `bokeh serve` and another running a flask app. I pull a session (which has some generic data in it), access the 'source' then I change the source based on who the user is etc.

The main issue that I am facing with this setup is that it works great for a small dataframe, but it somehow silently fails for a larger data set (not too large, 15k rows with 7 columns). What do you think is going on here?

So to summarize - Using @birdsarah's solution here works well for a smaller dataframe (5k rows with 7 columns), but fails with a larger one (15k rows with 7 columns). Any idea how I can diagnose?
